### PR TITLE
feat(consensus/validator): timestamp validation

### PR DESCRIPTION
**Pull request by sistemd**

### CODE REVIEW

File: crates/pathfinder/src/validator.rs [hunk 1]
```diff
@@ -101,7 +102,9 @@ impl ValidatorBlockInfoStage {
             )));
         }
 
-        // TODO(validator) validate block info (timestamp, gas prices)
+        validate_block_info_timestamp(block_info.height, block_info.timestamp, &main_storage)?;
+
+        // TODO(validator) validate gas prices
```
Comment (sistemd) [inline crates/pathfinder/src/validator.rs:107]: What exactly should we validate with regards to gas prices?

File: crates/pathfinder/src/validator.rs [hunk 2]
```diff
@@ -1408,4 +1481,137 @@ mod tests {
...
+        } else {
+            assert!(result.is_ok());
+        }
+    }
+
+    #[rstest]
+    #[case(BlockNumber::GENESIS)]
+    #[ignore = "TODO With deferred execution, not having a parent in the database is considered
+        valid when receiving proposal parts. We could also have deferred block info validation,
+        where we wait until parent is committed before validating timestamps."]
```
Comment (sistemd) [inline crates/pathfinder/src/validator.rs:1564]: WDYT?

File: crates/pathfinder/src/validator.rs (hunk 1 above)
Comment (kkovaacs) [reply crates/pathfinder/src/validator.rs:107]: Apollo is scraping L1 gas prices from Ethereum and L2 gas prices are determined based on that as far as I understand. We should find out the algorithm (average of the latest N L1 blocks?) and implement a scraper ourselves so that we can validate the gas prices.

(https://github.com/starkware-libs/sequencer/tree/main/crates/apollo_l1_gas_price)

File: crates/pathfinder/src/consensus/inner/p2p_task/handler_proptest.rs [hunk 3]
```diff
@@ -345,15 +345,17 @@ fn create_structurally_invalid_proposal(
...
-        .first()
-        .is_some_and(|first| first.is_proposal_init())
-        && proposal_parts
-            .get(1)
-            .is_some_and(|part| part.is_block_info())
-        && proposal_parts
-            .last()
-            .is_some_and(|last| last.is_proposal_fin())
+    match proposal_parts {
+        [] => true,
```
Comment (vbar) [inline crates/pathfinder/src/consensus/inner/p2p_task/handler_proptest.rs]: Why should an empty vector be considered well-ordered? It wasn't before, and looking at usage, it'll only make the caller panic...

Comment (sistemd) [reply crates/pathfinder/src/consensus/inner/p2p_task/handler_proptest.rs]: True, an empty proposal should not happen here, and should result in a panic. Handling it more explicitly now: https://github.com/eqlabs/pathfinder/pull/3159/commits/8f873b236e7fda2cbc9d99e43062bbe06e4432f3

File: crates/pathfinder/src/validator.rs (hunk 1 above)
Comment (sistemd) [reply crates/pathfinder/src/validator.rs:107]: @t00ts self-assigned this: https://github.com/eqlabs/pathfinder/issues/3160

File: crates/pathfinder/src/consensus/inner/p2p_task/handler_proptest.rs (hunk 3 above)
Comment (vbar) [reply crates/pathfinder/src/consensus/inner/p2p_task/handler_proptest.rs]: I've actually seen that panic when running proptests (didn't save the seed, though :-( ) - IMHO the mutated proposal could be empty...

Comment (vbar) [reply crates/pathfinder/src/consensus/inner/p2p_task/handler_proptest.rs]: here it is, against the latest version:
```
   thread 'consensus::inner::p2p_task::handler_proptest::test_handle_incoming_proposal_part' (612856) panicked at crates/pathfinder/src/consensus/inner/p2p_task/handler_proptest.rs:349:15:
    Proposal should not be empty
    note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

    thread 'consensus::inner::p2p_task::handler_proptest::test_handle_incoming_proposal_part' (612856) panicked at crates/pathfinder/src/consensus/inner/p2p_task/handler_proptest.rs:349:15:
    Proposal should not be empty

    thread 'consensus::inner::p2p_task::handler_proptest::test_handle_incoming_proposal_part' (612856) panicked at crates/pathfinder/src/consensus/inner/p2p_task/handler_proptest.rs:349:15:
    Proposal should not be empty

    thread 'consensus::inner::p2p_task::handler_proptest::test_handle_incoming_proposal_part' (612856) panicked at crates/pathfinder/src/consensus/inner/p2p_task/handler_proptest.rs:50:1:
    Test failed: Proposal should not be empty.
    minimal failing input: (proposal_type, seed) = (
        StructurallyInvalidExecutionFails,
        3944679451671559893,
    )
    	successes: 0
    	local rejects: 0
    	global rejects: 0
```

The seed diff is
```
--- a/crates/pathfinder/proptest-regressions/consensus/inner/p2p_task/handler_proptest.txt
+++ b/crates/pathfinder/proptest-regressions/consensus/inner/p2p_task/handler_proptest.txt
@@ -9,3 +9,4 @@ cc bd3644c258f2b87d94ad2fefacf8b776dc6dc858d32e40ae4e689eb7dd04901a # shrinks to
 cc 8773afd684ac46fd41aafb884b7679f57a4450fe3b297fd8a9efd0390d8959d8 # shrinks to (proposal_type, seed) = (StructurallyInvalidExecutionOk, 9181708935074874480)
 cc 102b2839e4f00df2f7d70c33871ab16a1a23bae6230551778c7335855a33d360 # shrinks to (proposal_type, seed) = (StructurallyInvalidExecutionOk, 10609718920510075896)
 cc b33e8ec389839f514689ca7e29df1e11a2254e917de4e3077485e97ddf747eb3 # shrinks to (proposal_type, seed) = (StructurallyInvalidExecutionFails, 8605337213757032975)
+cc cfdfe140297c5eae6faa431ea1cb80bbb184f735ba721b57f164e4f0d464ed38 # shrinks to (proposal_type, seed) = (StructurallyInvalidExecutionFails, 3944679451671559893)
```

Comment (sistemd) [reply crates/pathfinder/src/consensus/inner/p2p_task/handler_proptest.rs]: Thanks! https://github.com/eqlabs/pathfinder/pull/3159/commits/1b99bb6d660b45e355e8dfb4fd444b5b6fbe0502

Comment (vbar) [reply crates/pathfinder/src/consensus/inner/p2p_task/handler_proptest.rs]: Well, I don't know why the proptest couldn't test the empty sequence? OTOH I don't have a strong opinion about it - it's just an obscure edge case...

**Review by CHr15F0x — APPROVED** [review]
